keps/sig-testing/draft-20190119-20190118-breaking-apart-the-kubernetes-test-tarball.md

```diff
@@ -0,0 +1,176 @@
+
+## Proposal
+
+Instead of building and distributing a single `kubernetes-test.tar.gz` with all
+portable sources and compiled binaries for all platforms, produce many
+platform-specific tarballs, one for each platform defined in 
+[`KUBE_TEST_PLATFORMS`](https://github.com/kubernetes/kubernetes/blob/193f659a1cd454b93cbe1e7b1f13b77c21783461/hack/lib/golang.sh#L150-L160):
+* `kubernetes-test-linux-amd64.tar.gz` 
+* `kubernetes-test-linux-arm.tar.gz` 
+* `kubernetes-test-linux-arm64.tar.gz` 
+* `kubernetes-test-linux-s390x.tar.gz` 
+* `kubernetes-test-linux-ppc64le.tar.gz` 
+* `kubernetes-test-darwin-amd64.tar.gz` 
+* `kubernetes-test-windows-amd64.tar.gz` 
+
+### Internal structure of the test tarball
+At present, the Kubernetes test tarball has several components, all 
+rooted under a `kubernetes/` top-level directory.
+
+#### Binary artifacts
+The test binary artifacts are currently organized into directories divided by platform:
+  * `platforms/`
+    * `darwin/`, `linux/`, `windows/`
+      * `amd64/`, `arm/`, `arm64/`, `ppc64le/`, `s390x/`
+
+For comparison, the existing platform-specific tarballs
+(such as `kubernetes-client-linux-amd64.tar.gz`) place all binaries under
+a constant path with no platform information: `kubernetes/client/bin/kubectl`.
+
+Scripts (such as `cluster/get-kube-binaries.sh`) [extract these tarballs
+back into platform-specific directories](https://github.com/kubernetes/kubernetes/blob/193f659a1cd454b93cbe1e7b1f13b77c21783461/cluster/get-kube-binaries.sh#L143-L156)
+to support downloading multiple platforms into a single workspace.
+
+The test tarball should follow the lead of the other platform-specific tarballs
+and place the binaries under `test/bin`. We can then reuse the existing
+functionality already implemented for the other tarballs.
+
+#### Portable sources
+Portable sources are basically copied directly from the source tree:
+  * `test/e2e/testing-manifests/`
+  * `test/images/`
+  * `test/kubemark/`
+  * `hack/` ([partially](https://github.com/kubernetes/kubernetes/blob/193f659a1cd454b93cbe1e7b1f13b77c21783461/hack/lib/golang.sh#L193-L197))
+
+We have two options for these:
+1. Continue to distribute as a separate tarball, either `kubernetes-test.tar.gz`,
+   or possibly something like `kubernetes-test-portable.tar.gz`.
+  * This makes the distinction very clear vs. the binary artifacts
+  * There's already some precedent, such as the `kubernetes-manifest.tar.gz` tarball
+  * It slightly complicates downloading of test dependencies
+2. Duplicate these sources into each binary-specific tarball.
+  * Simplifies test dependency distribution
+  * Portable sources are small (as of v1.13.2, approximately 2.7MB uncompressed
+    or about 186KB compressed) so duplication isn't a huge worry
+  * Complicates extraction of tarballs with existing scripts, since they assume
+    everything is platform-specific
+
+It's not immediately obvious which of these is the better approach based on the
```

Comment:
do we know what tooling is pulling these tarballs besides get-kube / kubetest? 1) may break less things but otherwise might not be as preferable

Reply:
on cs.k8s.io I'm not finding much outside k/k. 

I should go digging through old issues/emails from the `kubernetes.tar.gz` work to refresh my memory on any issues encountered back then.

Reply:
I would like to be able to pull a subset of this - specifically e2e.test. I often use a private Jenkins instance running to test things before I submit PRs.

Reply:
IIRC the binaries are also pushed to GCS (outside the tarballs) when we stage a release, and for presubmit specifically this binary can be built by each of the build tools.

Reply:
we don't currently push any of the test binaries by themselves, though I guess we could if there's a strong desire for that. seems out of scope for this KEP, though.